sway-core/src/declaration_engine/declaration_engine.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{
+    concurrent_slab::ConcurrentSlab,
+    semantic_analysis::{
+        TypedImplTrait, TypedStructDeclaration, TypedTraitDeclaration, TypedTraitFn,
+    },
+    TypedFunctionDeclaration,
+};
+
+use super::{declaration_id::DeclarationId, declaration_wrapper::DeclarationWrapper};
+
+/// Used inside of type inference to store declarations.
+pub struct DeclarationEngine {
+    slab: ConcurrentSlab<DeclarationId, DeclarationWrapper>,
+    // *declaration_id -> vec of monomorphized copies
+    // where the declaration_id is the original declaration
+    monomorphized_copies: im::HashMap<usize, Vec<DeclarationId>>,
```

Comment:
Any reason to not use `HashMap<DeclarationId, ...>`?

Reply:
Just a regular HashMap would probably be better.